File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<pair<int,int>> adj[n];
        for(auto it: flights){
            adj[it[0]].push_back({it[1],it[2]});
        }
        queue<pair<int,pair<int,int>>>q;
        q.push({0,{src,0}});
        vector<int>dist(n,1e9);
        
        while(!q.empty()){
            auto p=q.front();
            q.pop();
            int stops=p.first;
            int node=p.second.first;
            int cost=p.second.second;
            if(stops>k) continue;
            for(auto it:adj[node]){

                int to=it.first;
                int weight=it.second;
                if(dist[to]>cost+weight){
                    dist[to]=cost+weight;
                    q.push({stops+1,{to,cost+weight}});
                }
            }
        }

        if(dist[dst]==1e9) return -1;
        return dist[dst];

    }
};

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        
        # src -> (dst, cost)
        graph = {}
        for edge in flights:
            graph[edge[0]] = graph.get(edge[0], [])
            graph[edge[0]].append((edge[1], edge[2]))

        # find cheapest flight between src and dst
        @cache
        def helper(src, dst, k):
            # print(src, dst, k)
            if k < 0:
                return math.inf

            ans = math.inf
            for neighbor in graph.get(src, []):
                # cheapest flight between neighbor and dst + cost of src and neighbor
                if neighbor[0] == dst:
                    ans = min(ans, neighbor[1])
                else:
                    ans = min(ans, helper(neighbor[0], dst, k - 1) + neighbor[1])

            return ans

        result = helper(src, dst, k)
        return -1 if result == math.inf else result